### PR TITLE
#3666 - Queue Monitoring - Enable Prometheus Metrics

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Enabled `/metrics` endpoint on `queue-consumers` to expose Prometheus metrics following [BC Gov docs](https://developer.gov.bc.ca/docs/default/component/platform-developer-docs/docs/app-monitoring/user-defined-monitoring/).
- Added [prom-client](https://github.com/siimon/prom-client) also following the BC Gov docs mentioned above and the one that seems largely adopted.
- Created a [Gauge metric](https://prometheus.io/docs/concepts/metric_types/#gauge) to capture the most recent summary of every active queue allowing to add to the metrics a summary of `active`, `completed`, `failed`, `delayed`, and 'waiting' jobs. This represents the same status observed in the Bull Board. The garage metrics rely on Redis queries and will always get the most updated values from Redis.
image
- Created a [Counter metric](https://prometheus.io/docs/concepts/metric_types/#counter) to capture all the local events triggered for a queue. This happens only in-memory and is captured every time the `metrics` endpoint is invoked. This metric is not needed to achieve queue monitoring but seems a great addition and useful data to support future analysis.
- Enable default [nodejs metrics](https://github.com/siimon/prom-client?tab=readme-ov-file#default-metrics) following also [Prometheus recommendations](https://prometheus.io/docs/instrumenting/writing_clientlibs/#standard-and-runtime-collectors).
- Both metrics are incremented/set using the labels `queueName`, `queueEvent`, and `queueType` to allow querying in Sysdig.

## Sysdig POCs

The Sysdig configurations are not final and were created to support the validation of the code in this PR but should not be considered final or part of the PR evaluation.

### Alerts generated for a queue with a failed job.

```
max(queue_job_counts_current_total{queueEvent="failed",kube_namespace_name="0c27fb-dev"}) by (queueName) > 0
```

image

image

### Same alerts are configured to be sent using an email channel.

image

## Sample Dashboard

The [Queues Overview](https://app.sysdigcloud.com/#/dashboards/419533?last=3600&scope=kubernetes.namespace.name%20%3D%20%3F%220c27fb-dev%22
) dashboard has some examples of data but should not be considered final or part of the PR evaluation.

image

_Note:_ the sysdig users and roles were updated and added to this PR, and it is already deployed to both tools environments. If time allows, further effort can be made to enhance the current process but any action beyond the user list update is not part of this PR.

### CODE REVIEW

File: sources/packages/backend/apps/queue-consumers/src/queues-metrics.module.module.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import { InjectLogger } from "@sims/utilities/logger";
+import { MetricsService } from "./services";
+
+@Global()
+@Module({
+  providers: [MetricsService],
+  exports: [MetricsService],
+})
+export class QueuesMetricsModule implements OnModuleInit {
+  constructor(private readonly metricsService: MetricsService) {}
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/queues-metrics.module.module.ts:11]: Do we require this un-used service at constructor?

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/queues-metrics.module.module.ts:11]: Got it now. The comment was added to the wrong file. I will remove it from the controller.

File: sources/packages/backend/apps/queue-consumers/src/services/metrics/metrics.service.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  /**
+   * Get the list of monitored queues.
+   * If not loaded yet, load it once.
+   * @returns list of monitored queues.
+   */
+  private async getMonitoredQueues(): Promise<MonitoredQueue[]> {
+    if (!this.monitoredQueueProviders) {
+      const queues = await this.queueService.queueConfigurationModel();
+      this.monitoredQueueProviders = queues
+        .filter((queueModel) => queueModel.isActive)
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/metrics/metrics.service.ts:140]: Are we not ignoring the FT queues here?

File: sources/packages/backend/apps/queue-consumers/src/services/metrics/metrics.service.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      labelNames: ["queueName", "queueEvent", "queueType"] as const,
+    });
+  }
+
+  /**
+   * Set global metrics configurations.
+   */
+  setGlobalMetricsConfigurations(): void {
+    register.setDefaultLabels({ app: DEFAULT_METRICS_APP_LABEL });
+    collectDefaultMetrics({ labels: { app: DEFAULT_METRICS_APP_LABEL } });
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/queue-consumers/src/services/metrics/metrics.service.ts:60]: for collect default metrics, is it not necesscary to set the register. 

```ts
collectDefaultMetrics({
      labels: { app: DEFAULT_METRICS_APP_LABEL, register },
    });
```

File: sources/packages/backend/apps/queue-consumers/src/services/metrics/metrics.service.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/metrics/metrics.service.ts:140]: The full-time toggle is not checked here. If the queues are not executing they will not produce any event.
The events will be associated but never triggered. I do not see any harm in it. Please let me know if further discussion would be required on this.

File: sources/packages/backend/apps/queue-consumers/src/services/metrics/metrics.service.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/metrics/metrics.service.ts:60]: Yes, there is no need to set the register unless using a custom one.

image
https://github.com/siimon/prom-client?tab=readme-ov-file#default-metrics

File: sources/packages/backend/apps/queue-consumers/src/services/metrics/metrics.service.ts (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/queue-consumers/src/services/metrics/metrics.service.ts:140]: Not required and I am good based on what we discussed.

**Review by dheepak-aot — APPROVED** [review]
Great work and amazing solution to share metrics to Sysdig. 

Thanks for doing the change. 👍

**Review by sh16011993 — APPROVED** [review]
Great Work @andrewsignori-aot 👍 Thank you for the PR walkthrough.

**Pull request merged**